### PR TITLE
Fix the token verify error

### PR DESCRIPTION
**Pull request by LinuxSuRen**
## How to verify?

1. Install KubeSphere with [branch remove-devops-ctrl](https://github.com/LinuxSuRen/kubesphere/tree/remove-devops-ctrl)
2. Install ks-devops, then update the image`surenpi/devops-apiserver:dev--43c24d5`
3. Install https://github.com/kubesphere-sigs/ks-devops-plugin

### CODE REVIEW

File: pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go [hunk 1]
```diff
@@ -16,14 +16,18 @@ func New() authenticator.Token {
 	return &tokenAuthenticator{}
 }
 
-func (a *tokenAuthenticator) AuthenticateToken(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+func (a *tokenAuthenticator) AuthenticateToken(ctx context.Context, token string) (response *authenticator.Response, ok bool, err error) {
 	issuer := jwt.NewTokenIssuer("", time.Second)
```
Comment (JohnNiang) [inline pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go:20]: IMO, the JWT secret should be set correctly.

Comment (JohnNiang) [reply pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go:20]: The documentation of `tokenAuthenticator` type should be rectified as well.
```
// tokenAuthenticator implements an anonymous auth
type tokenAuthenticator struct{}
```

File: pkg/apiserver/authentication/token/jwt.go [hunk 2]
```diff
@@ -45,22 +46,42 @@ type jwtTokenIssuer struct {
...
-	token, err := parser.Parse(tokenString, s.keyFunc)
-	if err != nil {
-		if token != nil {
-			mapClaims := token.Claims.(jwt.MapClaims)
 
-			return &user.DefaultInfo{
-				Name: mapClaims["sub"].(string),
-			}, "clm.TokenType", nil
+	// set up the default values
+	tokenType = "clm.TokenType"
```
Comment (JohnNiang) [inline pkg/apiserver/authentication/token/jwt.go:55]: Why is default token type `clm.TokenType`?

File: pkg/apiserver/authentication/token/jwt.go [hunk 3]
```diff
@@ -45,22 +46,42 @@ type jwtTokenIssuer struct {
...
+	userInfo = &user.DefaultInfo{}
+
+	var token *jwt.Token
+	// TODO consider to verify the token parse result
+	if token, _ = parser.Parse(tokenString, s.keyFunc); token != nil {
+		mapClaims := token.Claims.(jwt.MapClaims)
```
Comment (JohnNiang) [inline pkg/apiserver/authentication/token/jwt.go:61]: We should verify if the type convention is successful

File: pkg/apiserver/filters/requestinfo.go [hunk 4]
```diff
@@ -75,8 +75,9 @@ func WithRequestInfo(handler http.Handler, resolver request.RequestInfoResolver)
 }
 
 func injectToken(req *http.Request) (requestWithTokenContext *http.Request) {
-	authorization := req.Header.Get("Authorization")
+	authorization := req.Header.Get("X-Authorization")
 	token := strings.ReplaceAll(authorization, "Bearer ", "")
+	token = strings.ReplaceAll(token, "bearer ", "")
```
Comment (JohnNiang) [inline pkg/apiserver/filters/requestinfo.go:80]: Word `bearer` should be capital here.

File: pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go (hunk 1 above)
Comment (LinuxSuRen) [reply pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go:20]: > IMO, the JWT secret should be set correctly.

It's not in this case. As you can see, I use `VerifyWithoutClaimsValidation`. So, we don't need a real token here.

Comment (LinuxSuRen) [reply pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go:20]: > The documentation of `tokenAuthenticator` type should be rectified as well.
> 
> ```
> // tokenAuthenticator implements an anonymous auth
> type tokenAuthenticator struct{}
> ```

Sure.

File: pkg/apiserver/authentication/token/jwt.go (hunk 2 above)
Comment (LinuxSuRen) [reply pkg/apiserver/authentication/token/jwt.go:55]: I just copied the code from [KubeSphere code base](https://github.com/kubesphere/kubesphere). I prefer to keep it until we have enough time to refactor it.

File: pkg/apiserver/filters/requestinfo.go (hunk 4 above)
Comment (LinuxSuRen) [reply pkg/apiserver/filters/requestinfo.go:80]: See also line 79.

`token := strings.ReplaceAll(authorization, "Bearer ", "")`

I keep using `bearer` and `Bearer` for compatibility purposes.

File: pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go (hunk 1 above)
Comment (JohnNiang) [reply pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go:20]: Alright

File: pkg/apiserver/authentication/token/jwt.go (hunk 2 above)
Comment (JohnNiang) [reply pkg/apiserver/authentication/token/jwt.go:55]: Got it.

File: pkg/apiserver/filters/requestinfo.go (hunk 4 above)
Comment (JohnNiang) [reply pkg/apiserver/filters/requestinfo.go:80]: My bad.

Comment (LinuxSuRen) [reply pkg/apiserver/filters/requestinfo.go:80]: No worries.

File: pkg/apiserver/authentication/token/jwt.go (hunk 3 above)
Comment (LinuxSuRen) [reply pkg/apiserver/authentication/token/jwt.go:61]: Sure.

Comment (LinuxSuRen) [reply pkg/apiserver/authentication/token/jwt.go]: sure. please check https://github.com/kubesphere-sigs/ks-devops/pull/56/commits/8e702d811c2bcde6ac197638d5322dfbf370fb55

File: pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go (hunk 1 above)
Comment (LinuxSuRen) [reply pkg/apiserver/authentication/authenticators/bearertoken/bearertoken.go:20]: Sure, please check https://github.com/kubesphere-sigs/ks-devops/pull/56/commits/8e702d811c2bcde6ac197638d5322dfbf370fb55

File: pkg/apiserver/filters/requestinfo.go [hunk 5]
```diff
@@ -75,8 +75,9 @@ func WithRequestInfo(handler http.Handler, resolver request.RequestInfoResolver)
 }
 
 func injectToken(req *http.Request) (requestWithTokenContext *http.Request) {
-	authorization := req.Header.Get("Authorization")
+	authorization := req.Header.Get("X-Authorization")
```
Comment (JohnNiang) [inline pkg/apiserver/filters/requestinfo.go:78]: This kind of break change should be mentioned in README.

**Review by JohnNiang — COMMENTED** [review]
hi @LinuxSuRen , I left one comment here. please take a moment to have a look.

File: pkg/apiserver/filters/requestinfo.go (hunk 5 above)
Comment (LinuxSuRen) [reply pkg/apiserver/filters/requestinfo.go:78]: Of course

Comment (LinuxSuRen) [reply pkg/apiserver/filters/requestinfo.go:78]: I put this into a good-first-issue. https://github.com/kubesphere-sigs/ks-devops/issues/68

**Pull request merged**